wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/PluginMetaRetriever.java

```diff
@@ -0,0 +1,266 @@
+          format(
+              "Error occurred on retrieval of plugin %s. Error: %s",
+              id + ':' + version, e.getMessage()));
+    }
+  }
+
+  private void validateMeta(PluginMeta meta, String id, String version)
+      throws InfrastructureException {
+    requireNotNullNorEmpty(meta.getId(), CHE_PLUGIN_OBJECT_ERROR, id, version, "ID is missing.");
+    requireEqual(
+        id,
+        meta.getId(),
+        "Plugin id in attribute doesn't match plugin metadata. Plugin object seems broken.");
+    requireNotNullNorEmpty(
+        meta.getVersion(), CHE_PLUGIN_OBJECT_ERROR, id, version, "Version is missing.");
+    requireEqual(
+        version,
+        meta.getVersion(),
+        "Plugin version in workspace config attributes doesn't match plugin metadata. Plugin object seems broken.");
+    requireNotNullNorEmpty(
+        meta.getName(), CHE_PLUGIN_OBJECT_ERROR, id, version, "Name is missing.");
+    requireNotNullNorEmpty(
+        meta.getType(), CHE_PLUGIN_OBJECT_ERROR, id, version, "Type is missing.");
+    requireNotNullNorEmpty(meta.getUrl(), CHE_PLUGIN_OBJECT_ERROR, id, version, "URL is missing.");
+  }
+
+  @VisibleForTesting
+  protected <T> T getBody(URI uri, Class<T> clas) throws IOException {
+    HttpURLConnection httpURLConnection = null;
+    try {
+      httpURLConnection = (HttpURLConnection) uri.toURL().openConnection();
+
+      int responseCode = httpURLConnection.getResponseCode();
+      if (responseCode != 200) {
+        throw new IOException(
+            format(
+                "Can't get object by URI '%s'. Error: %s",
+                uri.toString(), getError(httpURLConnection)));
+      }
+
+      return parseYamlResponseStreamAndClose(httpURLConnection.getInputStream(), clas);
+    } finally {
+      if (httpURLConnection != null) {
+        httpURLConnection.disconnect();
+      }
+    }
+  }
+
+  private String getError(HttpURLConnection httpURLConnection) throws IOException {
+    try (InputStreamReader isr = new InputStreamReader(httpURLConnection.getInputStream())) {
+      return CharStreams.toString(isr);
+    }
+  }
+
+  protected <T> T parseYamlResponseStreamAndClose(InputStream inputStream, Class<T> clazz)
+      throws IOException {
+    try (InputStreamReader reader = new InputStreamReader(inputStream)) {
+      return YAML_PARSER.readValue(reader, clazz);
+    } catch (IOException e) {
+      throw new IOException(
+          "Internal server error. Unexpected response body received from Che plugin registry API."
+              + e.getLocalizedMessage(),
+          e);
+    }
+  }
+
+  @SuppressWarnings("SameParameterValue")
+  private void requireNotNullNorEmpty(String s, String error, String... errorArgs)
+      throws InfrastructureException {
+    if (s == null || s.isEmpty()) {
+      throw new InfrastructureException(format(error, (Object[]) errorArgs));
+    }
+  }
+
+  private void requireNotNullNorEmpty(Collection objects, String error, String... errorArgs)
```

Comment:
Four methods are unused. Do you think it makes sense to hold them here?

Reply:
ok